### PR TITLE
Sessionless in C#

### PR DESCRIPTION
**Pull request by Galacticai**
# Sessionless in kotlin
### classlib:
src/csharp/Sessionless
### Tests: (incomplete)
src/csharp/SessionlessTests

# Highlights:
- ### Util/
    Contains utilities for generating and converting keys... and more 
- ### Models/
    Abstract concepts
- ### Impl/
    Implementations of `models/` + non-abstract stuff

# Main focus:
- ### ..../Models/**ISessionless.cs**
- ### ..../Impl/**Sessionless.cs** (implementation of `ISessionless`)
- ### ..../Models/**IVault.cs** (The way to get/store key pairs)

# Notice:
Almost all classes, objects, and functions are documented using xaml docs which means that your IDE will show you the documentation while typing/hovering/etc...

# Class diagrams: 

# .../Models/**ISessionless**.cs

```mermaid
classDiagram
  class ISessionless {
    +Vault: IVault
    +GenerateUUID(): string
    +GenerateKeys(): KeyPairHex
    +GenerateKeysAsync(): Task<`KeyPairHex>
    +GetKeys(): KeypairHex?
    +Sign(string): MessageSignatureHex
    +Sign(string, string): MessageSignatureHex
    +Sign(string, ECPrivateKeyParameters): MessageSignatureHex
    +VerifySignature(SignedMessage): bool
    +VerifySignature(SignedMessage, string): bool
    +VerifySignature(SignedMessage, ECPublicKeyParameters): bool
    +Associate(SignedMessages[]): bool
  }
  class Sessionless {
    constructor(IVault)
  }
  ISessionless <-- Sessionless
```

# .../Models/**IVault**.cs

```mermaid
classDiagram
  class IVault {
    +Get(): KeyPairHex?
    +Save(KeyPairHex)
  }
  class Vault {
    constructor(Func<`KeyPairHex?>, Action<`KeyPairHex>)
  }

  IVault <-- Vault
```

# The rest:

### .../Impl/MessageSignature.cs

```mermaid
classDiagram
  class IMessageSignature { }
  class MessageSignatureInt {
    +R: BigInteger
    +S: BigInteger
    +constructor(BigInteger, BigInteger)
    +constructor(BigInteger[])
    +constructor(MessageSignatureHex)
    +toHex(): MessageSignatureHex
    +op_explicit(MessageSignatureInt): MessageSignatureHex
  }
  class MessageSignatureHex {
    +RHex: string
    +SHex: string
    +constructor(string, string)
    +constructor(MessageSignatureInt)
    +toInt(): MessageSignatureInt
    +op_explicit(MessageSignatureHex): MessageSignatureInt
  }

  IMessageSignature <-- MessageSignatureInt
  IMessageSignature <-- MessageSignatureHex
```

### .../Impl/...

```mermaid
classDiagram
  class SignedMessage {
    +Message: string
    +Signature: MessageSignatureHex
    +PublicKey: string
    +constructor(string, MessageSignatureHex, string)
  }

  class KeyPairHex {
    +PrivateKey: string
    +PublicKey: string
    +constructor(string, string)
  }
```

### .../Impl/Exceptions/...

```mermaid
classDiagram
  class KeyPairNotFoundException {
    constructor()
  }
  class HexFormatRequiredException {
    constructor(string)
  }
  Exception <-- KeyPairNotFoundException
  Exception <.. FormatException
  FormatException <-- HexFormatRequiredException
```

### CODE REVIEW

File: src/csharp/Sessionless/Impl/Sessionless.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        signer.Init(true, paramWithRandom);
+        // message string to keccak256 hash
+        byte[] messageHash = message.HashKeccak256();
+        // sign
+        BigInteger[] signature = signer.GenerateSignature(messageHash);
+        return new MessageSignatureInt(signature).ToHex();
+    }
+
+
+    public bool Verify(SignedMessage signedMessage) {
```
Comment (zach-planet-nine) [inline src/csharp/Sessionless/Impl/Sessionless.cs:72]: `VerifySignature` here

File: src/csharp/Sessionless/Impl/Sessionless.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        var publicHex = GetKeys()?.PublicKey
+            ?? throw new KeyPairNotFoundException();
+        if (publicHex != signedMessage.PublicKey) return false;
+        return Verify(signedMessage, publicHex);
+    }
+    public bool Verify(SignedMessage signedMessage, string publicKeyHex) {
```
Comment (zach-planet-nine) [inline src/csharp/Sessionless/Impl/Sessionless.cs:78]: `VerifySignature`

File: src/csharp/Sessionless/Impl/Sessionless.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        if (publicKeyHex != signedMessage.PublicKey) return false;
+        // public hex to bytes
+        var publicBytes = Hex.Decode(publicKeyHex);
+        // public bytes to key object
+        ECDomainParameters curve = KeyUtils.Defaults.DomainParameters;
+        ECPoint qPoint = curve.Curve.DecodePoint(publicBytes);
+        var publicKey = new ECPublicKeyParameters(qPoint, curve);
+        return Verify(signedMessage, publicKey);
+    }
+    public bool Verify(SignedMessage signedMessage, ECPublicKeyParameters publicKey) {
```
Comment (zach-planet-nine) [inline src/csharp/Sessionless/Impl/Sessionless.cs:91]: `VerifySignature`

File: src/csharp/Sessionless/Impl/Sessionless.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        // signature hex to bigint (and hex still included)
+        var signature = signedMessage.Signature.ToInt();
+        // message string to keccak256 hash
+        byte[] messageHash = signedMessage.Message.HashKeccak256();
```
Comment (zach-planet-nine) [inline src/csharp/Sessionless/Impl/Sessionless.cs:95]: we only do the first 32 bytes of the messageHash

**Review by zach-planet-nine — APPROVED** [review]
Looks good other than the naming. We'll need to get an example spun up. I have a unity game we could throw this in or we could spin up a new project.

File: src/csharp/Sessionless/Impl/Sessionless.cs (hunk 4 above)
Comment (Galacticai) [reply src/csharp/Sessionless/Impl/Sessionless.cs]: `HashKeccak256` is an extension method which produces 32 bytes
### Full method:
```csharp
    /// <summary> Get the hash of <paramref name="input"/> in keccak256 format </summary>
    internal static byte[] HashKeccak256(this string input) {
        // input to bytes
        byte[] bytes = Encoding.UTF8.GetBytes(input);
        //init digest
        KeccakDigest keccak256 = new(256);
        keccak256.BlockUpdate(bytes, 0, bytes.Length);
        byte[] hashBytes = new byte[keccak256.GetDigestSize()];
        // fill hash bytes
        keccak256.DoFinal(hashBytes, 0);
        return hashBytes;
    }
```

### The focus is:
```csharp
// Keccak-256 Algorithm
KeccakDigest keccak256 = new(256);

// size of digest (256 bits - 32bytes) as size of array
byte[] hashBytes = new byte[keccak256.GetDigestSize()];
```

**Pull request merged**